internal:Happy eyeballs

larry-safran opened the pull request:
Fix the Happy Eyeballs work (pr #10486) done by Tony on PickFirstLeafLoadBalancer and enhance tests.

larry-safran commented:
Rebased onto master.  @YifeiZhuang it is ready to review.

YifeiZhuang commented on core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:
```diff
@@ -225,15 +253,52 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
+            break;
+          }
+
+          boolean hasIdle = false;
```
I might miss something, could you point me to the section in grfc about the hasIdle/hasConnecting thing.
Also what is missing here is to count the TF times starting from the first pass based on grfc `Happy Eyeballs in the pick_first LB Policy
` section

> If the first pass completes without a successful connection attempt, ... We will count the number of connection failures, and when that number reaches the number of subchannels, we will request re-resolution;

larry-safran replied:
Updated to only do the refreshNameResolution when there have been sufficient connection failures.

larry-safran replied:
Happy Eyeballs should be the same as the petiole policies which had the rule that if any child is READY, then it is READY, if not but any child is CONNECTING then it is CONNECTING, if not but some child is IDLE then it is IDLE, otherwise it is TRANSIENT FAILURE.

YifeiZhuang replied:
Happy eyeballs is pick first not petiole policies. I don't think aggregated connectivity applies here this way. The aggregated connectivity processing here makes it unnecessarily complicated.

YifeiZhuang reviewed (commented):
Sending what I have. One question is that it looks not easy to make the happy eyeball logics controlled by a flag.

YifeiZhuang commented on core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:
```diff
@@ -324,6 +432,9 @@ private Subchannel createNewSubchannel(SocketAddress addr) {
         .addOption(HEALTH_CONSUMER_LISTENER_ARG_KEY, hcListener)
             .build());
+    if (subchannel == null) {
+      throw new IllegalArgumentException("Was not able to create subchannel for " + addr);
```
IIRC LB should never throw. Otherwise it cause channel panic.

larry-safran replied:
I can return a null, and then in requestConnection try the next address.

larry-safran commented on core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:
```diff
@@ -85,62 +93,68 @@ public Status acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
     final List<EquivalentAddressGroup> newImmutableAddressGroups =
         Collections.unmodifiableList(new ArrayList<>(servers));
 
+    boolean reuseReadyChannel = false;
```
done

larry-safran commented on core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:
```diff
@@ -85,62 +93,68 @@ public Status acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
       }
     }
 
+    if (reuseReadyChannel) {
```
done

larry-safran commented on core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:
```diff
@@ -179,6 +194,19 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
     if (newState == SHUTDOWN) {
       return;
     }
+
```
I don't follow how the subchannelData check is covering the passed in state value not being SHUTDOWN.  Can you clarify?

YifeiZhuang replied:
we already talked offline that when rawConnectivityState is READY, there is only one active subchannel remained, others are all shutdown. And L191 ignores callback from other subchannels different from the current one.

larry-safran commented on core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:
```diff
@@ -291,29 +357,71 @@ private void shutdownRemaining(SubchannelData activeSubchannelData) {
+        subchannels.get(currentAddress).updateState(CONNECTING);
+        scheduleNextConnection();
+        break;
+      case CONNECTING:
```
This is in `requestConnection`, not in `processSubchannelState`.   We have already made the decision that we are requesting a connection and are just trying to find the appropriate subchannel to request it on.

YifeiZhuang replied:
`requestConnection` can be from both. Upon address update, or as a result of `processSubchannelState`.
grfc made it clear that when we requesting a connection, this subchannel may NOT be in IDLE. 
> If the subchannel is in state CONNECTING, we do not need to actually request a connection, but we will treat it as if we did. If it is not the last subchannel in the list, we will start the timer; if it is the last subchannel in the list, we will wait for the attempt to complete.

So here we should not increment the index, instead we should restart the timer if it is not the last address.

larry-safran commented on core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:
```diff
@@ -291,29 +357,71 @@ private void shutdownRemaining(SubchannelData activeSubchannelData) {
         addressIndex.increment();
         requestConnection();
+        break;
+      case READY: // Since we aren't doing anything there won't be a state change
```
done

larry-safran commented:
@YifeiZhuang Addressed all of the comments including incorporating a flag to enable the logic.  Mostly, disabling is just not scheduling the try of the next subchannel on timeout.

YifeiZhuang commented on core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:
```diff
@@ -225,15 +244,40 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
                 new Picker(PickResult.withError(stateInfo.getStatus())));
+            break;
+          }
+        } else if (!addressIndex.isValid()) {
```
I feel this else-if block is hard to follow and needs cleanup. Isn't this enough:
```
else if (!addressIndex.isValid() && enableHappyEyeball && ++counter) {
  refresh name resolution.
} 
```

larry-safran commented on core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:
```diff
@@ -291,29 +343,92 @@ private void shutdownRemaining(SubchannelData activeSubchannelData) {
+      subchannel = subchannels.containsKey(currentAddress)
+          ? subchannels.get(currentAddress).getSubchannel()
+          : createNewSubchannel(currentAddress);
+    } while (subchannel == null && addressIndex.isValid());
```
Done.  Updated helper.createSubchannel to mention that null is not allowed and createNewSubchannel to throw an exception instead of returning null.

larry-safran commented on core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:
```diff
@@ -291,29 +343,92 @@ private void shutdownRemaining(SubchannelData activeSubchannelData) {
+        break;
+      case CONNECTING:
+        if (enableHappyEyeballs) {
+          if (!addressIndex.isAtEnd()) {
```
Done

larry-safran commented on core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:
```diff
@@ -324,19 +439,18 @@ private Subchannel createNewSubchannel(SocketAddress addr) {
             new EquivalentAddressGroup(addr)))
         .addOption(HEALTH_CONSUMER_LISTENER_ARG_KEY, hcListener)
             .build());
+    if (subchannel == null) {
```
If the helper is created correctly.  I would prefer not to trust something that we don't completely control and do the check.  However, I am now throwing an exception rather than returning null.

larry-safran commented on core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:
```diff
@@ -55,11 +57,20 @@
   private final Helper helper;
   private final Map<SocketAddress, SubchannelData> subchannels = new HashMap<>();
   private Index addressIndex;
+  private int numTfSinceAccept = 0;
```
Changed the logic to only start counting after the first pass is complete and added a test to make sure that we are following the grfc.

YifeiZhuang replied:
I would correct a wrong statement I mentioned above. Both acceptResolvedAddress and processSubchannelState (see [callback](https://github.com/grpc/grpc-java/blob/f01c110f7788d79ba87e5f37c82a320f8c35f1a3/core/src/main/java/io/grpc/internal/InternalSubchannel.java#L73) in InternalSubchannel) are in the same syncContext in managedChannelImpl, although they might from different threads. So there is no thread safety issue. And the `AddressIndex` can just be local without synchronized.

larry-safran replied:
Done

YifeiZhuang commented on core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:
```diff
@@ -211,32 +231,45 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
+        // If at end and: not doing happy eyeballs, just got to the end, or we have had enough TF
+        // since we got to the end, then refresh name resolution
+        if (!addressIndex.isValid()
+            && (!enableHappyEyeballs || numTf.get() >= addressIndex.size() || numTf.get() == 0)) {
```
I see that you combined a few cases here. It's not trivial to verify and hard to maintain. 
For example, it looks wrong in the case of `enableHappyEyeballs==false`, because name resolution is refreshed every time when addressIndex is invalid and we have TF. 
Also I think increment the counter and the action should be together instead of scatter around. What about move L198-L200 down, and what about call `helper.refreshNameresolution` twice, once for the first time it goes invalid, then every time the counter > address size.

YifeiZhuang commented on core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:
```diff
@@ -85,54 +100,57 @@ public Status acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
     }
+
+    // Since we have a new set of addresses, we are again at first pass
+    numTf.getAndSet(0);
```
I think remove these two lines. Because it is much simpler to count the numTF after first pass, instead of start from address update. That is easier to reason about. We have all the conditions to tell it is first pass so we do not need extra variable. See the code snippet in the later comments.
When we have address update, we reset the index already, (only case is READY connection and address still in the new list, when the READY connection goes IDLE, it will reset the index, then it comes to the same situation), so when the first time index goes to invalid when we iterating the list, we can tell it is first pass complete. 
In case that when we receive the exact same set of addresses and they are all in TF already, it will miss the first pass but it is fine because the it is name resolution is already refreshed before update.

larry-safran commented on core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:
```diff
@@ -211,32 +230,49 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
             && subchannels.get(addressIndex.getCurrentAddress()).getSubchannel() == subchannel) {
-          addressIndex.increment();
-          requestConnection();
+          if (addressIndex.increment()) {
```
There are 2 other cases where we hit the end  of addressIndex and for both of them we don't know that a connection attempt has completed:

- the happy eyeballs scheduled task's deadline being exceeded (`StartNextConnection.run`)
- `RequestConnectionPicker.pickSubchannel`

YifeiZhuang replied:
1. You are right. I see that the "first pass complete" check I proposed is not sufficient in the case where they are incremented somewhere else.
One concern I have is that in those cases where the first pass is done happened in other places instead of `processSubchannelState`, we are checking `firstPass` at the next subchannel failure, which is too late, worse case is that the transport comes back until it times out.  I think the way C++ is doing is more accurate, https://github.com/grpc/grpc/blob/6a4b5ccea362dd47562c698dd2aa5a8aead41a6a/src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc#L1150 what do you think?
2. This is minor but you could have let the numTF only increment when first pass is done. But either way.
3. Add `enableHappyEyeballs` condition.

larry-safran replied:
1. We need to check when the connection attempt completes which is when it calls processSubchannelState, so I believe that is the appropriate way to do it.  I've updated the logic to actually look at the subchannel states to determine that they have all completed.
2. I believe that the new logic using passComplete() does this
3. The old PickFirstLoadBalancer calls refreshNameResolution on every TF it receives [code](https://github.com/grpc/grpc-java/blob/be74bbdfe0d8a799e7fc32155bb0772371ab8b4c/core/src/main/java/io/grpc/internal/PickFirstLoadBalancer.java#L116-L118).  I agree with an earlier comment you made that name resolution should be refreshed like in the happy eyeballs case only at end of first pass and after address count worth of TF status reports, though @ejona86 may disagree.

YifeiZhuang replied:
1.   When `requestConnection` from `pick()` or immediate after `acceptResolvedAddress` reaches the last address, it won't call back `processSubchannelState`,  e.g. when all subchannel in TF status already.  I believe the behaviour won't be a huge disaster but still different.
3. The old PickFirstLoadBalancer only has one subchannel but the iteration logic is inside one subchannel, so it calls refreshNameResolution every TF actually means every iteration of all addresses. It looks before your happy eyeballs change it only refreshNameResolution once, which seems a bug. I'm ok with your change this way unless Eric has different views.

YifeiZhuang replied:
I think this still remains an issue though:

> When requestConnection from pick() or immediately after acceptResolvedAddress reaches the last address in its iteration, `processSubchannelState` does not come back, e.g. when all subchannel in TF status already. This would cause the refreshNameResolution delayed.

If you have different opinion, let's engage Eric in the thread. Otherwise at least add comment to fixit.

larry-safran replied:
I created a couple of tests `happy_eyeballs_connection_results_happen_after_get_to_end` and `happy_eyeballs_pick_pushes_index_over_end` which try these use cases and they don't have any issues.

ejona86 commented on core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:
```diff
@@ -211,32 +230,49 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
+        // (just got to the end), then refresh name resolution which may call
+        // acceptResolvedAddresses which will reset addressIndex, but may not.
+        if (!addressIndex.isValid()) {
+          numTf.getAndIncrement();
```
It looks like `numTF` doesn't need to be atomic. And that's good, because these three calls to adjust it wouldn't be thread-safe. We should make it a normal integer.

larry-safran replied:
Changed to a normal int.

ejona86 commented on core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:
```diff
@@ -55,11 +58,21 @@
+  private ScheduledHandle scheduleConnectionTask;
   private ConnectivityState rawConnectivityState = IDLE;
   private ConnectivityState concludedState = IDLE;
+  private final boolean enableHappyEyeballs =
```
static?

larry-safran replied:
I wasn't able to think of a clean way to set it in tests if it was static.

YifeiZhuang commented on core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:
```diff
@@ -211,32 +228,48 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
+          //    * Have had enough TF reported since we completed first pass
+          //    * Just completed the first pass
+          if (++numTf >= addressIndex.size() || firstPass) {
+            firstPass = false; // idempotent so okay to call when false
```
Nit: I know what you meant by saying `idempotent` but technically the word is a more strict terminology. Just remove the comment would be clear.

YifeiZhuang commented on core/src/test/java/io/grpc/internal/PickFirstLeafLoadBalancerTest.java:
```diff
@@ -110,46 +125,53 @@ public void uncaughtException(Thread t, Throwable e) {
   private PickSubchannelArgs mockArgs;
 
+  @BeforeClass
+  public static void setUpClass() {
```
I didn't see you reset the property @AfterClass, so other tests in the entire grpc test will always see happyeyeballs enabled. However I think we expect them to test on happy eyeballs not enabled case.
More importantly, the test suite here should verify both enabled/disabled cases (use Parameterized), or at least have the original tests do the disabled cases, and have your new tests enable the variable. I didn't realize last time I saw.

larry-safran replied:
Changed to setting in `setup` and added an `@after` method to restore it.

larry-safran commented on core/src/test/java/io/grpc/internal/PickFirstLeafLoadBalancerTest.java:
```diff
@@ -110,46 +121,55 @@ public void uncaughtException(Thread t, Throwable e) {
   @Before
   public void setUp() {
-    for (int i = 1; i < 5; i++) {
+    System.setProperty(PickFirstLeafLoadBalancer.GRPC_EXPERIMENTAL_XDS_DUALSTACK_ENDPOINTS, "TRUE");
```
Done.

YifeiZhuang commented on core/src/test/java/io/grpc/internal/PickFirstLeafLoadBalancerTest.java:
```diff
@@ -2057,15 +2076,264 @@ public void ready_then_transient_failure_again() {
+
+    // Put the LB into TF
+    listeners[0].onSubchannelState(ConnectivityStateInfo.forTransientFailure(error));
+    listeners[1].onSubchannelState(ConnectivityStateInfo.forTransientFailure(error));
```
One example is here, say we have 3 subchannels(TF, TF, TF). Index points to invalid.
Then we have address update with exactly the same addresses. So we reset FirstPass and Index = 0. 
We start do requestConnection() which iterates all 3 addresses. 
After that we should have a `refreshNameResolution` but current behaviour won't, because that logic in only in the listener, which is not sufficient. Right? I understand this may complicate the logic, so if we do not have a simple solution we should comment.

larry-safran replied:
Since the subchannels are in TF, each of them is retrying their connections independently of the load balancer.  We reset firstPass to true, so as soon as one of those retries gets a TF response we'll call refreshNameResolution.

YifeiZhuang replied:
Talked offline about the (potentially delayed, but purposely) `refreshNameResolution` behaviour when all subchannels are in TF upon address update, or part of the subchannels are in TF upon address update. This looks fine to me.

YifeiZhuang reviewed (approved):
LGTM
